### PR TITLE
Large tests and runnable example

### PR DESCRIPTION
**Pull request by marcin-krolik**
- added large tests
- added runnable example

@intelsdi-x/plugin-maintainers

### CODE REVIEW

**Comment by marcin-krolik** [comment]
PR #10 has to be merged **before** this one in order to enable runnable example and large test to download plugin binary from S3

File: examples/task/mock-load.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+. "${__dir}/../../scripts/common.sh"
+
+# ensure PLUGIN_PATH is set
+TMPDIR=${TMPDIR:-"/tmp"}
+PLUGIN_PATH=${PLUGIN_PATH:-"${TMPDIR}/snap/plugins"}
+mkdir -p $PLUGIN_PATH
+
+_info "downloading plugins"
+(cd $PLUGIN_PATH && curl -sSO http://snap.ci.snap-telemetry.io/snap/master/latest/snap-plugin-publisher-mock-file && chmod 755 snap-plugin-publisher-mock-file)
+(cd $PLUGIN_PATH && curl -sSO http://snap.ci.snap-telemetry.io/snap/master/latest/snap-plugin-processor-passthru && chmod 755 snap-plugin-processor-passthru)
```
Comment (IRCody) [inline examples/task/mock-load.sh]: Since passthru isn't loaded below, does it need to be downloaded?

File: examples/task/mock-load.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ for i in `seq 1 10`; do
+            _info "try ${i}"
+             if [[ -f /usr/local/bin/snapctl && -f /usr/local/bin/snapd ]];
+                then
+                    _info "loading plugins"
+                    snapctl plugin load "${PLUGIN_PATH}/snap-plugin-publisher-mock-file"
+                    snapctl plugin load "${PLUGIN_PATH}/snap-plugin-collector-load"
+
+                    _info "creating and starting a task"
+                    snapctl task create -t "${__dir}/task-mem.json"
```
Comment (IRCody) [inline examples/task/mock-load.sh]: This file doesn't seem to exist AFAICT.

File: scripts/test/large.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        self.assertEqual(len(plugins), 2, "Plugins available {} expected {}".format(len(plugins), 2))
+
+        # check for snapd errors
+        self.assertEqual(len(self.binaries.snapd.errors), 0, "Errors found during snapd execution")
+
+    def tearDown(self):
+        log.info("stopping snapd")
+        self.binaries.snapd.stop()
+        if self.binaries.snapd.isAlive():
+            log.warn("snapd thread did not died")
```
Comment (IRCody) [inline scripts/test/large.py]: "snapd thread did not die"

**Comment by nanliu** [comment]
@marcin-krolik, I've been playing with large tests in serverspec:
https://github.com/intelsdi-x/snap-plugin-collector-ethtool/pull/20

If we are interested in using python, could we use a test framework like testinfra?
https://testinfra.readthedocs.io/en/latest/

Either serverspec or testinfra would allow us to write much more succinct tests and be platform agnostic. As our test matrix gets more complex there's a lot of value to leverage a framework so we can focus on test instead of getting everything ready to test.

We should discuss this in next weeks sync up.

**Comment by marcin-krolik** [comment]
@IRCody 
>I think it would be beneficial to get consensus on that before going to far down different roads for large tests across multiple plugins.

I agree.

@nanliu I'm not familiar with neither serverspec nor testinfra. My try with wrapper module and unittest is a proposition of simple framework for large test. I wanted to keep docker image small as possible, thus I didn't want to add any additional dependencies and used standard modules + my own implementation. I'm opened for discussion and I totally agree with your point we need some kind of test framework for large tests.

File: examples/task/mock-load.sh (hunk 1 above)
Comment (marcin-krolik) [inline examples/task/mock-load.sh]: It should be loaded. Added loading of passthru

File: examples/task/mock-load.sh (hunk 2 above)
Comment (marcin-krolik) [inline examples/task/mock-load.sh]: Fixed to proper task manifest file

File: scripts/test/large.py (hunk 3 above)
Comment (marcin-krolik) [inline scripts/test/large.py]: Fixed typo

**Comment by marcin-krolik** [comment]
Large test will fail at the end with error `AssertionError: Errors found during snapd execution` due to [bug](https://github.com/intelsdi-x/snap/issues/1229)

**Comment by marcintao** [comment]
LGTM

File: .gitmodules [hunk 4]
```diff
@@ -0,0 +1,3 @@
+[submodule "scripts/test/pytest"]
+	path = scripts/test/pytest
+	url = https://github.com/marcin-krolik/snap-pytest
```
Comment (marcin-krolik) [inline .gitmodules]: @IRCody Yes, I will prepare pip package.

**Comment by marcin-krolik** [comment]
@IRCody Submodule is removed. Additionally I updated paths to S3, added glide.lock, script for running large tests on local k8s (minikube).

**Comment by marcin-krolik** [comment]
@IRCody removed .gitmodules

File: examples/tasks/mock-load.sh [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+(cd $PLUGIN_PATH && curl -sSO http://snap.ci.snap-telemetry.io/snap/latest_build/linux/x86_64/snap-plugin-publisher-mock-file && chmod 755 snap-plugin-publisher-mock-file)
+(cd $PLUGIN_PATH && curl -sSO http://snap.ci.snap-telemetry.io/snap/latest_build/linux/x86_64/snap-plugin-processor-passthru && chmod 755 snap-plugin-processor-passthru)
+(cd $PLUGIN_PATH && curl -sSO http://snap.ci.snap-telemetry.io/plugins/snap-plugin-collector-load/latest_build/linux/x86_64/snap-plugin-collector-load && chmod 755 snap-plugin-collector-load)
+
+SNAP_FLAG=0
+
+# this block will wait check if snapctl and snapd are loaded before the plugins are loaded and the task is started
+ for i in `seq 1 10`; do
+            _info "try ${i}"
+             if [[ -f /usr/local/bin/snapctl && -f /usr/local/bin/snapd ]];
```
Comment (mjbrender) [inline examples/tasks/mock-load.sh]: Does this need to use the renamed binaries?

File: examples/tasks/README.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        `snap-plugin-publisher-mock-file` and starts the task
+- [run-mock-load.sh](run-mock-load.sh)
+    - The example is launched with this script     
+- [task-load.json](task-load.json)
+    - Snap task definition
+- [.setup.sh](.setup.sh)
+    - Verifies dependencies and starts the containers.  It's called 
+    by [run-mock-load.sh](run-mock-load.sh).
+- [docker-compose.yml](docker-compose.yml)
+    - A docker compose file which defines "runner" container where snapd
```
Comment (mjbrender) [inline examples/tasks/README.md]: binaries renamed from snapctl/snapd to snaptel/d

**Pull request merged**